File: classic150/Solution0045.cpp
#include <vector>
using namespace std;
// 给定一个长度为 n 的 0 索引整数数组 nums。初始位置在下标 0。

// 每个元素 nums[i] 表示从索引 i 向后跳转的最大长度。换句话说，如果你在索引 i 处，你可以跳转到任意 (i + j) 处：

// 0 <= j <= nums[i] 且
// i + j < n
// 返回到达 n - 1 的最小跳跃次数。测试用例保证可以到达 n - 1。

// 示例 1:

// 输入: nums = [2,3,1,1,4]
// 输出: 2
// 解释: 跳到最后一个位置的最小跳跃数是 2。
//      从下标为 0 跳到下标为 1 的位置，跳 1 步，然后跳 3 步到达数组的最后一个位置。
// 示例 2:

// 输入: nums = [2,3,0,1,4]
// 输出: 2

// 提示:

// 1 <= nums.length <= 104
// 0 <= nums[i] <= 1000
// 题目保证可以到达 n - 1
class Solution
{
public:
    int jump(vector<int> &nums)
    {
        if (nums.size() == 1)
            return 0;
        int count = 0;
        int i = 0;
        while (i < nums.size())
        {
            int nextPos = 0;
            int maxReach = 0;
            for (int j = 1; j <= nums[i]; j++)
            {
                if (i + j >= nums.size() - 1) // 能直接跳到终点
                    return count + 1;
                if (i + j + nums[i + j] > maxReach)
                {
                    maxReach = i + j + nums[i + j];
                    nextPos = i + j;
                }
            }
            i = nextPos;
            count++;
        }
        return count;
    }
};

// 从结尾开始跳，每次看谁能跳到这里，选下标最小的
class Solution2
{
public:
    int jump(vector<int> &nums)
    {
        int pos = nums.size() - 1;
        int count = 0;
        while (pos > 0)
        {
            for (int i = 0; i < pos; i++)
            {
                if (i + nums[i] >= pos)
                {
                    pos = i;
                    count++;
                }
            }
        }
        return count;
    }
};

class Solution3
{
public:
    int jump(vector<int> &nums)
    {
        int count = 0;
        int end = 0;
        int maxReach = 0; // 在[end,maxReach中试探]，每次end取上一个maxreach的值，当遍历到了end，说明必须跳了
        for (int i = 0; i < nums.size(); i++)
        {
            maxReach = max(maxReach, i + nums[i]);
            if (i == end)
            {
                end = maxReach;
                count++;
            }
        }
        return count;
    }
};

File: anew/Solution0045.java
package anew;

// 给定一个长度为 n 的 0 索引整数数组 nums。初始位置在下标 0。

// 每个元素 nums[i] 表示从索引 i 向后跳转的最大长度。换句话说，如果你在索引 i 处，你可以跳转到任意 (i + j) 处：

// 0 <= j <= nums[i] 且
// i + j < n
// 返回到达 n - 1 的最小跳跃次数。测试用例保证可以到达 n - 1

public class Solution0045 {
    public int jump(int[] nums) {
        int n = nums.length;
        int[][] dp = new int[n][2];
        dp[0][0] = 0;
        dp[0][1] = nums[0];
        for (int i = 1; i < n; i++) {
            dp[i][0] = Integer.MAX_VALUE;
            for (int j = 0; j < i; j++) {
                if (i <= dp[j][1]) {
                    dp[i][0] = Math.min(dp[i][0], dp[j][0] + 1);
                    dp[i][1] = Math.max(dp[i][1], i + nums[i]);
                }
            }
        }
        return dp[n - 1][0];
    }

    // dp[i][0]是非递减的，所以可以用一个指针j来找到第一个可以跳到i的下标
    public int jump2(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];
        dp[0] = 0;

        int j = 0;
        for (int i = 1; i < n; i++) {
            while (j + nums[j] < i) {
                j++;
            }
            dp[i] = dp[j] + 1;
        }
        return dp[n - 1];
    }

    public int jump3(int[] nums) {
        int n = nums.length;
        int jump = 0;
        int curMaxReach = 0;
        int nextMaxReach = 0;
        for (int i = 0; i < n - 1; i++) {
            nextMaxReach = Math.max(nextMaxReach, i + nums[i]);
            if (i == curMaxReach) {
                jump++;
                curMaxReach = nextMaxReach;
            }
        }
        return jump;
    }

    public static void main(String[] args) {
        Solution0045 solution = new Solution0045();
        int[] nums = { 2, 3, 1, 1, 4 };
        System.out.println(solution.jump(nums));
        System.out.println(solution.jump2(nums));
        System.out.println(solution.jump3(nums));
        int[] nums2 = { 2, 3, 0, 1, 4 };
        System.out.println(solution.jump(nums2));
        System.out.println(solution.jump2(nums2));
        System.out.println(solution.jump3(nums2));
    }
}
